quinn-proto/src/tests/mod.rs

```diff
@@ -1177,6 +1177,39 @@ fn keep_alive() {
     }
 }
 
+#[test]
+fn cid_retirement() {
+    let _guard = subscribe();
+    let mut pair = Pair::default();
+    let (client_ch, server_ch) = pair.connect();
+
+    // Server retires current active remote CIDs
+    pair.server_conn_mut(server_ch).mock_retire_cid(1, 1);
+    pair.drive();
+    // Any unexpected behavior may trigger TransportError::CONNECTION_ID_LIMIT_ERROR
+    assert!(!pair.client_conn_mut(client_ch).is_closed());
+    assert!(!pair.server_conn_mut(server_ch).is_closed());
+    assert_matches!(pair.client_conn_mut(client_ch).active_rem_cid_seq(), 1);
+
+    use crate::cid_queue::CidQueue;
+    use crate::LOC_CID_COUNT;
+    let mut active_cid_num = CidQueue::LEN as u64 + 1;
+    active_cid_num = active_cid_num.min(LOC_CID_COUNT);
+
+    let _next_retire_prior_to = active_cid_num + 1;
+    pair.client_conn_mut(client_ch).ping();
+    // Server retires all valid remote CIDs
+    pair.server_conn_mut(server_ch)
+        .mock_retire_cid(_next_retire_prior_to, 1);
+    pair.drive();
+    assert!(!pair.client_conn_mut(client_ch).is_closed());
+    assert!(!pair.server_conn_mut(server_ch).is_closed());
+    assert_matches!(
+        pair.client_conn_mut(client_ch).active_rem_cid_seq(),
+        _next_retire_prior_to
+    );
```

Comment:
When i checkout commit () (fix bug in NEW_CONNECTION_ID but not fixing in RETIRE_CONNECTION_ID [here](https://github.com/quinn-rs/quinn/pull/871/files#diff-423ed266e9aaf5fcae0fa8b85ca382f495528c18cd7c890cef1cf438221f2ef6R2363-R2379)), I got the following output

```
Oct 16 10:56:19.798 TRACE server:send{space=Data pn=3}: quinn_proto::connection: NEW_CONNECTION_ID sequence=5 id=eaa580934a8f5476
Oct 16 10:56:19.799 TRACE server: quinn_proto::connection: sending 62 byte datagram
Oct 16 10:56:19.799 TRACE client: quinn_proto::connection: got Data packet (62 bytes) from [::1]:4433 using id f08a8aa4490cf931
Oct 16 10:56:19.799 TRACE client:recv{space=Data pn=3}: quinn_proto::connection: authenticated
Oct 16 10:56:19.799 TRACE client:recv{space=Data pn=3}:frame{ty=NEW_CONNECTION_ID}: quinn_proto::connection: sequence=5 id=eaa580934a8f5476 retire_prior_to=1
Oct 16 10:56:19.799 TRACE client:recv{space=Data pn=3}:frame{ty=NEW_CONNECTION_ID}: quinn_proto::connection: switching to remote CID 1: 2af6dbf3da6f1e1f
Oct 16 10:56:19.799 TRACE client:send{space=Data pn=3}: quinn_proto::connection: ACK
Oct 16 10:56:19.799 TRACE client:send{space=Data pn=3}: quinn_proto::connection: RETIRE_CONNECTION_ID sequence=0
Oct 16 10:56:19.800 TRACE client: quinn_proto::connection: sending 36 byte datagram
Oct 16 10:56:19.800 TRACE server: quinn_proto::connection: got Data packet (36 bytes) from [::1]:44433 using id 2af6dbf3da6f1e1f
Oct 16 10:56:19.801 TRACE server:recv{space=Data pn=3}: quinn_proto::connection: authenticated
Oct 16 10:56:19.801 TRACE server:send{space=Data pn=4}: quinn_proto::connection: ACK
Oct 16 10:56:19.801 TRACE server: quinn_proto::connection: sending 34 byte datagram
Oct 16 10:56:19.801 TRACE server: quinn_proto::endpoint: peer retired CID 0: 782fdbd230ae62b1
Oct 16 10:56:19.801 TRACE client: quinn_proto::connection: got Data packet (34 bytes) from [::1]:4433 using id f08a8aa4490cf931
Oct 16 10:56:19.801 TRACE client:recv{space=Data pn=4}: quinn_proto::connection: authenticated
Oct 16 10:56:19.802 TRACE server:send{space=Data pn=5}: quinn_proto::connection: ACK
Oct 16 10:56:19.802 TRACE server:send{space=Data pn=5}: quinn_proto::connection: NEW_CONNECTION_ID sequence=6 id=67d47302296af768
Oct 16 10:56:19.803 TRACE server: quinn_proto::connection: sending 62 byte datagram
Oct 16 10:56:19.803 TRACE client: quinn_proto::connection: got Data packet (62 bytes) from [::1]:4433 using id f08a8aa4490cf931
Oct 16 10:56:19.803 TRACE client:recv{space=Data pn=5}: quinn_proto::connection: authenticated
Oct 16 10:56:19.804 TRACE client:recv{space=Data pn=5}:frame{ty=NEW_CONNECTION_ID}: quinn_proto::connection: sequence=6 id=67d47302296af768 retire_prior_to=1
Oct 16 10:56:19.804 DEBUG client: quinn_proto::connection: closing connection due to transport error: the number of connection IDs provided by the peer exceeds the advertised active_connection_id_limit
Oct 16 10:56:19.805 TRACE client: quinn_proto::connection: connection closed
Oct 16 10:56:19.805 TRACE client:send{space=Data pn=4}: quinn_proto::connection: sending CONNECTION_CLOSE
Oct 16 10:56:19.805 TRACE client: quinn_proto::connection: sending 30 byte datagram
Oct 16 10:56:19.806 TRACE server: quinn_proto::connection: got Data packet (30 bytes) from [::1]:44433 using id 2af6dbf3da6f1e1f
Oct 16 10:56:19.806 TRACE server:recv{space=Data pn=4}: quinn_proto::connection: authenticated
Oct 16 10:56:19.806 TRACE server: quinn_proto::connection: connection closed
Oct 16 10:56:19.806 TRACE server:send{space=Data pn=6}: quinn_proto::connection: sending CONNECTION_CLOSE
Oct 16 10:56:19.807 TRACE server: quinn_proto::connection: sending 30 byte datagram
Oct 16 10:56:19.807 TRACE client: quinn_proto::connection: got Data packet (30 bytes) from [::1]:4433 using id f08a8aa4490cf931
Oct 16 10:56:19.807 TRACE client:recv{space=Data pn=6}: quinn_proto::connection: draining
Oct 16 10:56:19.808 TRACE quinn_proto::tests::util: advancing to Instant { t: 306065761809108 } for client
Oct 16 10:56:19.808 TRACE client: quinn_proto::connection: timeout timer=Close
Oct 16 10:56:19.808 TRACE server: quinn_proto::connection: timeout timer=Close
thread 'tests::cid_retirement' panicked at 'assertion failed: !pair.client_conn_mut(client_ch).is_closed()', quinn-proto/src/tests/mod.rs:1190:5
```

You can see server continue to send
`Oct 16 10:56:19.802 TRACE server:send{space=Data pn=5}: quinn_proto::connection: NEW_CONNECTION_ID sequence=6 id=67d47302296af768
` to client and then triggers the limitexceed error on client side - matches with what i described previously.